### ISSUE TITLE
nhml benchmarking doesn't make sense

### ISSUE THREAD

**Issue by leftishness** [nicehash/NiceHashMiner#407]
I'm not a programmer. So, I'm sorry I can't give more detailed info about this issue but I've been trying to benchmark 1 rig for days. There are some algos that will just stall and the benchmark will not advance. 

How is there not already some built-in timer that restarts the process automatically when it's obviously gone too long!? I realize it's difficult to set a maximum limit on how long a given algo benchmark should take with different hardware but it's _definitely_ somewhere between 0 < x < 10 minutes on standard.

Also, when benchmarks fail, why is it not automatically retrying them until it succeeds? What user is going to just say, "Oh the benchmark failed, I guess I'll just make less money and not use that algo." The benchmarking aspect of this software doesn't seem very well thought out. Why do I have to interact with it so much just to benchmark my cards? It's really silly.

Also, it would be really nice if, once benchmarking is complete, the benchmark data were portable between version upgrades or even wiping and reinstalling Windows. That way, you only ever have to benchmark once for a given set of hardware and can just back up the data on a USB drive. Once Windows and NHML are reinstalled, just drop the folder with the benchmark data into a specific directory and you're mining. This would save a lot of time, given how often Windows' instability requires a system wipe.

**Comment by C1BERUS** [comment]
it's already there, just keep config folder on NHML to your backup, copy it to new version of NHML. done, u dont need to benchmark again when upgrading to new version.

**Comment by Horsetoothless** [comment]
I got them all benchmarked some may require actual mining to take place and extended precise time entered in settings screen wish mining window display during benchmarking for valid checking.
Add Neoscrypt to build please.

**Comment by b0b5h4rp13** [comment]
This reminds me of when I did not have enough power for the GPU's I was tring to run. Check you power cables, power supply wattage

**Comment by SergeySypalo** [comment]
Btw, is it possible to auto-update benchmarking results?
I mean, when new version of NiceHash Miner Legacy is released, I'm just extracting it with overwrite to my existing folder with NHML. Last time when upgraded from 1.8.3 to 1.8.4-Pre2 it's asked me to relauch to download new miners. 
But the problem is that benchmark results are old and it's not obvious and timeconsuming to manually rerun all benchmarks for all card after each update. Ideally it would be to update benchmarked result if higher result is achieved. 
So load all results into memory to do not deal with filesystem and compare from time to time (almost all miners have an API that can be queried for current status, which can report current mining speed). And if new version of any 3rd party miner is able to achieve higher speed, just update text config file.
Same for initial run. Just allow user to run miner without prebenchmarking and just grab the results for different algos on the way. Surely it won't mine preffered one because of lack of results, but there are plently of room for improvement.
For example pre-populate values from generic database, and update if result is higher. Also if I have 6 same cards, doesn't make much sense to check every one. Just compare clocks and if they are the same - assume that they can achieve the same speed.

**Comment by DillonN** [comment]
@UnderwearWarrior  there is already a timeout built into the benchmarking process, which are you finding that never end? Some of the benchmarks take a very long time (e.g. xmr-stak-cpu takes 12min on a 6 core CPU). There are certainly some improvements I'd like to make to the benchmark process, but right now they are lower priority since NHML is designed for more advanced users. 

As @C1BERUS said the benchmarks are already portable (the files are all in the configs folder). Updating by overwriting into the current folder is not recommended as you will not get updates to miner files, see [here](https://github.com/NiceHash/NiceHashMinerLegacy/wiki/Getting-started#upgrading)

Having NHML benchmark algorithms as it mines has been brought up before, but not as a method of benchmarking updated miner files. The problem is once the miner file has been updated in a new NHML version, the switching may never pick it if the older speed is unprofitable. There isn't a reliable way to instruct NHML to use updated miners since that would require it forcing the updated benchmark speed through a database or something. Since every machine is different this may result in users mining suboptimal algorithms once they hit start for the first time

I think going forward there should be three improvements that would solve most of the issues:

* Have a per-benchmark progress indicator (instead of just an overall one). This would give users info on whether the bench has hanged or is just long, or if it was cut short.

* Have NHML store miner versions in benchmark files, so on upgrade any updated miners will be flagged as needs benchmark (currently users must read the release notes to see if one has updated, then rebenchmark it manually)

* Have a passive benchmark that uses mining stats to flesh out a more accurate benchmark speed (would be user configurable as to how long it lasts).
